src/tensorflow/tensorflow-rnnlm-lib.cc

```diff
@@ -0,0 +1,361 @@
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+
+#include <utility>
+#include <fstream>
+
+#include "tensorflow/core/public/session.h"
+#include "tensorflow/core/platform/env.h"
+#include "tensorflow/core/protobuf/meta_graph.pb.h"
+
+#include "tensorflow/tensorflow-rnnlm-lib.h"
+#include "util/stl-utils.h"
+#include "util/text-utils.h"
+
+namespace kaldi {
+using std::ifstream;
+using tf_rnnlm::KaldiTfRnnlmWrapper;
+using tf_rnnlm::TfRnnlmDeterministicFst;
+using tensorflow::Status;
+
+// read a unigram count file of the OOSs and generate extra OOS costs for words
+void SetUnkPenalties(const string &filename,
+                     const fst::SymbolTable& fst_word_symbols,
+                     std::vector<float> *out) {
+  if (filename == "")
+    return;
+  out->resize(fst_word_symbols.NumSymbols(), 0);  // default is 0
+  ifstream ifile(filename.c_str());
+  string word;
+  float count, total_count = 0;
+  while (ifile >> word >> count) {
+    int id = fst_word_symbols.Find(word);
+    KALDI_ASSERT(id != fst::SymbolTable::kNoSymbol);
+    (*out)[id] = count;
+    total_count += count;
+  }
+
+  for (int i = 0; i < out->size(); i++) {
+    if ((*out)[i] != 0) {
+      (*out)[i] = log ((*out)[i] / total_count);
+    }
+  }
+}
+
+// Read tensorflow checkpoint files
+void KaldiTfRnnlmWrapper::ReadTfModel(const std::string &tf_model_path,
+                                      int32 num_jobs) {
+  string graph_path = tf_model_path + ".meta";
+
+  tensorflow::SessionOptions session_options;
+  session_options.config.set_intra_op_parallelism_threads(num_jobs); // limit parallelism within jobs
+  session_options.config.set_inter_op_parallelism_threads(num_jobs); // limit parallelism within jobs
+
+  Status status = tensorflow::NewSession(session_options,
+                                         &session_);
+  if (!status.ok()) {
+    KALDI_ERR << status.ToString();
+  }
+
+  tensorflow::MetaGraphDef graph_def;
+  status = tensorflow::ReadBinaryProto(tensorflow::Env::Default(), graph_path,
+                                       &graph_def);
+  if (!status.ok()) {
+    KALDI_ERR << status.ToString();
+  }
+
+  // Add the graph to the session
+  status = session_->Create(graph_def.graph_def());
+  if (!status.ok()) {
+    KALDI_ERR << status.ToString();
+  }
+
+  Tensor checkpointPathTensor(tensorflow::DT_STRING, tensorflow::TensorShape());
+  checkpointPathTensor.scalar<std::string>()() = tf_model_path;
+  
+  status = session_->Run(
+      {{ graph_def.saver_def().filename_tensor_name(), checkpointPathTensor },},
```

Comment:
It's a [list initialization](http://en.cppreference.com/w/cpp/language/list_initialization), the type of the initialized object is inferred from the called function signature - seems valid C++11, although the use of the comma operator (the one before the last) makes this look very strange (not sure about whether that comma is needed though).